File: pgpainless-core/src/main/java/org/pgpainless/key/protection/BaseSecretKeyRingProtector.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.protection;

import org.bouncycastle.openpgp.PGPException;
import org.bouncycastle.openpgp.operator.PBESecretKeyDecryptor;
import org.bouncycastle.openpgp.operator.PBESecretKeyEncryptor;
import org.pgpainless.implementation.ImplementationFactory;
import org.pgpainless.key.protection.passphrase_provider.SecretKeyPassphraseProvider;
import org.pgpainless.util.Passphrase;

import javax.annotation.Nullable;

/**
 * Basic {@link SecretKeyRingProtector} implementation that respects the users {@link KeyRingProtectionSettings} when
 * encrypting keys.
 */
public class BaseSecretKeyRingProtector implements SecretKeyRingProtector {

    private final SecretKeyPassphraseProvider passphraseProvider;
    private final KeyRingProtectionSettings protectionSettings;

    /**
     * Constructor that uses the given {@link SecretKeyPassphraseProvider} to retrieve passphrases and PGPainless'
     * default {@link KeyRingProtectionSettings}.
     *
     * @param passphraseProvider provider for passphrases
     */
    public BaseSecretKeyRingProtector(SecretKeyPassphraseProvider passphraseProvider) {
        this(passphraseProvider, KeyRingProtectionSettings.secureDefaultSettings());
    }

    /**
     * Constructor that uses the given {@link SecretKeyPassphraseProvider} and {@link KeyRingProtectionSettings}.
     *
     * @param passphraseProvider provider for passphrases
     * @param protectionSettings protection settings
     */
    public BaseSecretKeyRingProtector(SecretKeyPassphraseProvider passphraseProvider, KeyRingProtectionSettings protectionSettings) {
        this.passphraseProvider = passphraseProvider;
        this.protectionSettings = protectionSettings;
    }

    @Override
    public boolean hasPassphraseFor(long keyId) {
        return passphraseProvider.hasPassphrase(keyId);
    }

    @Override
    @Nullable
    public PBESecretKeyDecryptor getDecryptor(long keyId) throws PGPException {
        Passphrase passphrase = passphraseProvider.getPassphraseFor(keyId);
        return passphrase == null || passphrase.isEmpty() ? null :
                ImplementationFactory.getInstance().getPBESecretKeyDecryptor(passphrase);
    }

    @Override
    @Nullable
    public PBESecretKeyEncryptor getEncryptor(long keyId) throws PGPException {
        Passphrase passphrase = passphraseProvider.getPassphraseFor(keyId);
        return passphrase == null || passphrase.isEmpty() ? null :
                ImplementationFactory.getInstance().getPBESecretKeyEncryptor(
                        protectionSettings.getEncryptionAlgorithm(),
                        protectionSettings.getHashAlgorithm(),
                        protectionSettings.getS2kCount(),
                        passphrase);
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/key/protection/BaseSecretKeyRingProtector.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.protection

import org.bouncycastle.bcpg.KeyIdentifier
import org.bouncycastle.openpgp.PGPPublicKey
import org.bouncycastle.openpgp.api.OpenPGPImplementation
import org.bouncycastle.openpgp.api.OpenPGPKey
import org.bouncycastle.openpgp.operator.PBESecretKeyDecryptor
import org.bouncycastle.openpgp.operator.PBESecretKeyEncryptor
import org.pgpainless.key.protection.passphrase_provider.SecretKeyPassphraseProvider

/**
 * Basic [SecretKeyRingProtector] implementation that respects the users [KeyRingProtectionSettings]
 * when encrypting keys.
 */
open class BaseSecretKeyRingProtector(
    private val passphraseProvider: SecretKeyPassphraseProvider,
    private val protectionSettings: KeyRingProtectionSettings
) : SecretKeyRingProtector {

    constructor(
        passphraseProvider: SecretKeyPassphraseProvider
    ) : this(passphraseProvider, KeyRingProtectionSettings.secureDefaultSettings())

    override fun hasPassphraseFor(keyIdentifier: KeyIdentifier): Boolean {
        return passphraseProvider.hasPassphrase(keyIdentifier)
    }

    override fun getDecryptor(keyId: Long): PBESecretKeyDecryptor? =
        getDecryptor(KeyIdentifier(keyId))

    override fun getDecryptor(keyIdentifier: KeyIdentifier): PBESecretKeyDecryptor? =
        passphraseProvider.getPassphraseFor(keyIdentifier)?.let {
            if (it.isEmpty) null
            else
                OpenPGPImplementation.getInstance()
                    .pbeSecretKeyDecryptorBuilderProvider()
                    .provide()
                    .build(it.getChars())
        }

    override fun getEncryptor(key: PGPPublicKey): PBESecretKeyEncryptor? {
        return passphraseProvider.getPassphraseFor(key.keyIdentifier)?.let {
            if (it.isEmpty) null
            else
                OpenPGPImplementation.getInstance()
                    .pbeSecretKeyEncryptorFactory(
                        false,
                        protectionSettings.encryptionAlgorithm.algorithmId,
                        protectionSettings.s2kCount)
                    .build(it.getChars(), key.publicKeyPacket)
        }
    }

    override fun getKeyPassword(p0: OpenPGPKey.OpenPGPSecretKey): CharArray? {
        return passphraseProvider.getPassphraseFor(p0.keyIdentifier)?.getChars()
    }
}
